File: LeetCode/src/dp/PerfectSquares.java
package dp;

public class PerfectSquares {
    public int numSquares(int n) {
        int[] dp = new int[n + 1];

        for (int j = 0; j < dp.length; j++) {
            dp[j] = Integer.MAX_VALUE;
        }

        dp[0] = 0;

        for (int i = 1; i * i <= n; i++) {
            for (int j = i * i; j <= n; j++) {
                dp[j] = Math.min(dp[j], dp[j - i * i] + 1);
            }
        }

        return dp[n];
    }

    public static void main(String[] args) {
        PerfectSquares solution = new PerfectSquares();

        // Test case
        int n = 12;
        System.out.println(solution.numSquares(n)); // Output: 3 (12 = 4 + 4 + 4)
    }
}

File: LeetCode_Python/dp/knapsack/PerfectSquares.py
class PerfectSquares:
    def numSquares(self, n: int) -> int:
        dp = [float("inf")] * (n + 1)
        dp[0] = 0
        squares = [i * i for i in range(1, int(n**0.5) + 1)]

        for s in squares:
            for i in range(s, n + 1):
                dp[i] = min(dp[i], dp[i - s] + 1)

        return dp[n]


# Time Complexity: O(n * sqrt(n)), where n is the target number
# Space Complexity: O(n), for the dp array

if __name__ == "__main__":
    solution = PerfectSquares()

    print("Expected Output: 3")
    print("Actual Output:", solution.numSquares(12))

    print("Expected Output: 2")
    print("Actual Output:", solution.numSquares(13))
